### DISCUSSION TITLE
Perform a teardown immediately after a session scoped setup fixture fails.

### DISCUSSION THREAD

**Discussion by EugeneBrod** [pytest-dev/pytest#12420] [Q&A] [answered]
Hi there.

I have 2 **session scoped** setup fixtures:
1. create device fixture # known good
2. configure device fixture # may raise exception

I'd like to perform an immediate teardown if the setup fails or if the call fails. The teardown would need the object created from fixture 1.

```
import pytest

# this hook implements a suggestion found here:
# https://docs.pytest.org/en/7.1.x/example/simple.html#making-test-result-information-available-in-fixtures
@pytest.hookimpl(hookwrapper=True)
def pytest_runtest_makereport(item):
    outcome = yield
    report = outcome.get_result()
    setattr(item, "rep_" + report.when, report) 


# known good fixture
@pytest.fixture(scope="session")
def create_device():
    device = 1
    yield device


# this fixture can fail
@pytest.fixture(scope="session")
def configure_device(create_device):
    device = create_device
    raise
    yield device


@pytest.fixture(autouse=True)
def teardown(create_device, request):
    yield
    device = create_device
    print("Entered teardown")

    # if setup or call failed
    if request.node.rep_setup.failed or request.node.rep_call.failed:
        print("Dumping device memory")


@pytest.fixture(autouse=True, scope="session")
def teardown(create_device, request):
    yield
    device = create_device
    print("Entered session scoped teardown")

    # if setup or call failed
    if request.node.testsfailed > 0:
        print("Dumping device memory")


def test_something(configure_device):
    device = configure_device
    print(device)
    pass
```

This approach works well when the failure occurs in the test call. But if the test setup fails, the teardown is deferred.

Looking for suggestions. Thanks in advance.

**Comment by The-Compiler** [comment]
I'd expect this to work off-hand. Could you show a [self-contained, complete example](http://www.sscce.org/), that's [formatted correctly](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks) so that someone can actually copy and run it?

**Reply by EugeneBrod** [reply, marked as answer]
**Problem** 
I wanted to create crashlogs immediately after failure in setup or test call. I wanted to generate crashlogs once for every fixture, whether session or function scoped. To generate crashlogs I needed the result of a reliable "base" fixture which is run first during test execution.

**Solution**
Define a "SessionControlBlock" class, which stores the result of the "base" fixture and a few methods for generating crashlogs. The class is assigned to the pytest's "session" object.

I then implemented the pytest_fixture_setup hook to 
1. save the result of the "base" fixture to the SessionControlBlock
2. determine if a fixture raised an exception, generate a crashlog if it did

The pytest_make_report hook was used to determine if the test failed during the "call" phase, generate a crashlog if it did.
